File: src/privacy/policy.py
"""
Signing Policy Engine

Evaluates whether a transaction should be signed based on:
- Mode-specific policy constraints
- Proof validity (private mode)
- Transfer limits and destination allowlists
- Replay protection

Each mode has different requirements:
  PUBLIC:  Standard validation (amount, destination, format)
  PRIVATE: Full ZK proof chain + standard validation + proof binding
"""

import enum
import time
from dataclasses import dataclass, field
from typing import List, Optional, Set

from src.zk.types import (
    ProofBundle,
    TransactionMode,
    TransferEnvelope,
)


class PolicyCheckResult(enum.Enum):
    """Result of a single policy check."""
    PASS = "pass"
    FAIL = "fail"
    SKIP = "skip"


@dataclass
class PolicyCheck:
    """A single policy check with name, result, and details."""
    name: str
    result: PolicyCheckResult
    detail: str = ""


@dataclass
class PolicyEvaluation:
    """Complete result of policy evaluation for a transaction."""
    approved: bool
    mode: TransactionMode
    checks: List[PolicyCheck] = field(default_factory=list)
    reason: str = ""
    timestamp: float = 0.0

    def __post_init__(self):
        if self.timestamp == 0.0:
            self.timestamp = time.time()

    @property
    def failed_checks(self) -> List[PolicyCheck]:
        return [c for c in self.checks if c.result == PolicyCheckResult.FAIL]

    def display(self) -> str:
        lines = [
            f"Policy Evaluation: {'APPROVED' if self.approved else 'REJECTED'}",
            f"Mode: {self.mode.value}",
        ]
        for check in self.checks:
            icon = "✓" if check.result == PolicyCheckResult.PASS else (
                "✗" if check.result == PolicyCheckResult.FAIL else "–"
            )
            lines.append(f"  {icon} {check.name}: {check.detail}")
        if self.reason:
            lines.append(f"Reason: {self.reason}")
        return "\n".join(lines)


class SigningPolicyEngine:
    """Evaluates signing policy for a transfer envelope.

    This is a lightweight policy layer on top of ZkProofEngine.validate_envelope().
    It wraps the TransferEnvelope type from types.py directly.
    """

    def __init__(
        self,
        max_transfer_lamports: int = 10_000_000_000,
        allowed_destinations: Optional[Set[str]] = None,
        require_destination_allowlist: bool = False,
    ):
        self.max_transfer_lamports = max_transfer_lamports
        self.allowed_destinations: Set[str] = allowed_destinations or set()
        self.require_destination_allowlist = require_destination_allowlist
        self._seen_nonces: Set[str] = set()

    def set_max_transfer(self, lamports: int):
        self.max_transfer_lamports = lamports

    def add_allowed_destination(self, address: str):
        self.allowed_destinations.add(address)

    def remove_allowed_destination(self, address: str):
        self.allowed_destinations.discard(address)

    def evaluate(self, envelope: TransferEnvelope) -> PolicyEvaluation:
        """Run all policy checks for a transfer envelope."""
        mode = envelope.mode
        checks: List[PolicyCheck] = []

        # 1. Mode validity
        checks.append(PolicyCheck(
            name="mode_valid",
            result=PolicyCheckResult.PASS,
            detail=f"Mode: {mode.value}",
        ))

        # 2. Envelope structure
        checks.append(self._check_structure(envelope))

        # 3. Amount within limits
        checks.append(self._check_amount(envelope))

        # 4. Destination check
        checks.append(self._check_destination(envelope))

        # 5. Replay protection
        checks.append(self._check_replay(envelope))

        # 6. Private-mode-only checks
        if mode == TransactionMode.PRIVATE:
            if envelope.proof_bundle is None:
                checks.append(PolicyCheck(
                    name="proof_bundle_present",
                    result=PolicyCheckResult.FAIL,
                    detail="Private mode requires a proof bundle",
                ))
            else:
                checks.append(PolicyCheck(
                    name="proof_bundle_present",
                    result=PolicyCheckResult.PASS,
                    detail="Proof bundle provided",
                ))
                checks.extend(self._check_proof_completeness(envelope.proof_bundle))
                checks.append(self._check_proof_binding(envelope))
        else:
            checks.append(PolicyCheck(
                name="proof_bundle_present",
                result=PolicyCheckResult.SKIP,
                detail="Not required for public mode",
            ))

        failed = [c for c in checks if c.result == PolicyCheckResult.FAIL]
        approved = len(failed) == 0
        reason = "; ".join(f"{c.name}: {c.detail}" for c in failed) if not approved else ""

        nonce = envelope.transaction.nonce
        if approved and nonce:
            self._seen_nonces.add(nonce)

        return PolicyEvaluation(approved=approved, mode=mode, checks=checks, reason=reason)

    # ── Individual Checks ──────────────────────────

    def _check_structure(self, envelope: TransferEnvelope) -> PolicyCheck:
        issues = []
        tx = envelope.transaction
        if not tx.from_pubkey:
            issues.append("missing sender")
        if not tx.to_pubkey:
            issues.append("missing recipient")
        if tx.amount_lamports < 0:
            issues.append("negative amount")
        if issues:
            return PolicyCheck(name="structure", result=PolicyCheckResult.FAIL,
                               detail=f"Invalid: {', '.join(issues)}")
        return PolicyCheck(name="structure", result=PolicyCheckResult.PASS,
                           detail="Envelope structure valid")

    def _check_amount(self, envelope: TransferEnvelope) -> PolicyCheck:
        amount = envelope.transaction.amount_lamports
        if amount > self.max_transfer_lamports:
            return PolicyCheck(name="amount_limit", result=PolicyCheckResult.FAIL,
                               detail=f"Amount {amount} exceeds limit {self.max_transfer_lamports}")
        return PolicyCheck(name="amount_limit", result=PolicyCheckResult.PASS,
                           detail=f"Amount {amount} within limit {self.max_transfer_lamports}")

    def _check_destination(self, envelope: TransferEnvelope) -> PolicyCheck:
        recipient = envelope.transaction.to_pubkey
        if not self.require_destination_allowlist:
            return PolicyCheck(name="destination", result=PolicyCheckResult.PASS,
                               detail="Destination allowlist not enforced")
        if not self.allowed_destinations:
            return PolicyCheck(name="destination", result=PolicyCheckResult.FAIL,
                               detail="Destination allowlist is enforced but empty")
        if recipient in self.allowed_destinations:
            return PolicyCheck(name="destination", result=PolicyCheckResult.PASS,
                               detail=f"Destination {recipient[:8]}… is allowed")
        return PolicyCheck(name="destination", result=PolicyCheckResult.FAIL,
                           detail=f"Destination {recipient[:8]}… not in allowlist")

    def _check_replay(self, envelope: TransferEnvelope) -> PolicyCheck:
        nonce = envelope.transaction.nonce
        if not nonce:
            return PolicyCheck(name="replay", result=PolicyCheckResult.FAIL,
                               detail="Missing nonce")
        if nonce in self._seen_nonces:
            return PolicyCheck(name="replay", result=PolicyCheckResult.FAIL,
                               detail=f"Nonce '{nonce[:16]}…' already used")
        return PolicyCheck(name="replay", result=PolicyCheckResult.PASS,
                           detail="Nonce is fresh")

    def _check_proof_completeness(self, bundle: ProofBundle) -> List[PolicyCheck]:
        checks = []
        if bundle.ownership_proof is not None:
            checks.append(PolicyCheck(name="ownership_proof", result=PolicyCheckResult.PASS,
                                      detail="Ownership proof present"))
        else:
            checks.append(PolicyCheck(name="ownership_proof", result=PolicyCheckResult.FAIL,
                                      detail="Missing ownership proof"))
        if bundle.range_proof is not None:
            checks.append(PolicyCheck(name="range_proof", result=PolicyCheckResult.PASS,
                                      detail="Range proof present"))
        else:
            checks.append(PolicyCheck(name="range_proof", result=PolicyCheckResult.SKIP,
                                      detail="Range proof not included"))
        if bundle.policy_proofs:
            checks.append(PolicyCheck(name="policy_proofs", result=PolicyCheckResult.PASS,
                                      detail=f"{len(bundle.policy_proofs)} policy proof(s)"))
        else:
            checks.append(PolicyCheck(name="policy_proofs", result=PolicyCheckResult.SKIP,
                                      detail="No policy proofs"))
        return checks

    def _check_proof_binding(self, envelope: TransferEnvelope) -> PolicyCheck:
        bundle = envelope.proof_bundle
        if bundle is None or not bundle.binding:
            return PolicyCheck(name="proof_binding", result=PolicyCheckResult.FAIL,
                               detail="No proof binding")
        return PolicyCheck(name="proof_binding", result=PolicyCheckResult.PASS,
                           detail="Proof binding present")

File: coldstar_zk/src/policy.rs
//! Signing Policy Engine
//!
//! Enforces signing policies for both public and private transactions.
//!
//! # Public Mode Policies
//! - Address validation
//! - Amount validation
//! - Fee sanity check
//!
//! # Private Mode Policies (in addition to public)
//! - All required proofs must be present
//! - All proofs must be valid
//! - Proof binding must be valid
//! - Nonce must be fresh (replay protection)
//! - Mode integrity (no mode switching after proof creation)

use std::collections::HashSet;

use crate::binding;
use crate::envelope;
use crate::error::ZkError;
use crate::proofs::{ownership, range};
use crate::types::{
    SigningSummary, TransactionMode, TransferEnvelope, VerificationCheck, VerificationResult,
};

/// Maximum transaction age in seconds (1 hour)
/// Reserved for production use — not enforced in MVP.
#[allow(dead_code)]
const MAX_TX_AGE_SECS: i64 = 3600;

/// Signing policy engine.
pub struct PolicyEngine {
    /// Set of previously seen nonces (replay protection)
    seen_nonces: HashSet<String>,
    /// Maximum allowed transfer in lamports (0 = no limit)
    max_transfer_lamports: u64,
    /// Allowed destination addresses (empty = allow all)
    allowed_destinations: HashSet<String>,
}

impl PolicyEngine {
    /// Create a new policy engine with default settings.
    pub fn new() -> Self {
        PolicyEngine {
            seen_nonces: HashSet::new(),
            max_transfer_lamports: 0, // no limit by default
            allowed_destinations: HashSet::new(),
        }
    }

    /// Set the maximum allowed transfer amount.
    pub fn set_max_transfer(&mut self, max_lamports: u64) {
        self.max_transfer_lamports = max_lamports;
    }

    /// Add an allowed destination address.
    pub fn add_allowed_destination(&mut self, address: String) {
        self.allowed_destinations.insert(address);
    }

    /// Validate a transfer envelope and produce a verification result + signing summary.
    ///
    /// This is the main entry point for the offline signer's policy checks.
    ///
    /// # Public Mode
    /// Runs standard policy checks only.
    ///
    /// # Private Mode
    /// Runs standard policy checks + proof verification + binding verification.
    ///
    /// # Returns
    /// (VerificationResult, SigningSummary) — the summary is what gets displayed
    /// to the user before signing.
    pub fn validate_envelope(
        &mut self,
        envelope: &TransferEnvelope,
    ) -> Result<(VerificationResult, SigningSummary), ZkError> {
        let mut checks = Vec::new();

        // === Structural Checks ===

        // Check 1: Envelope integrity
        let integrity_ok = envelope::verify_envelope_integrity(envelope).is_ok();
        checks.push(VerificationCheck {
            name: "Envelope integrity (HMAC)".to_string(),
            passed: integrity_ok,
            detail: if integrity_ok {
                "HMAC verification passed".to_string()
            } else {
                "HMAC verification FAILED — envelope may have been tampered with".to_string()
            },
        });

        // Check 2: Envelope structure
        let structure_ok = envelope::validate_envelope_structure(envelope).is_ok();
        checks.push(VerificationCheck {
            name: "Envelope structure".to_string(),
            passed: structure_ok,
            detail: if structure_ok {
                "Mode and proof bundle consistency verified".to_string()
            } else {
                "Structural validation failed".to_string()
            },
        });

        // === Standard Policy Checks (both modes) ===

        // Check 3: Amount sanity
        let amount_ok = envelope.transaction.amount_lamports > 0;
        checks.push(VerificationCheck {
            name: "Amount validation".to_string(),
            passed: amount_ok,
            detail: format!("{} lamports", envelope.transaction.amount_lamports),
        });

        // Check 4: Max transfer limit
        let limit_ok = self.max_transfer_lamports == 0
            || envelope.transaction.amount_lamports <= self.max_transfer_lamports;
        checks.push(VerificationCheck {
            name: "Transfer limit".to_string(),
            passed: limit_ok,
            detail: if self.max_transfer_lamports == 0 {
                "No limit configured".to_string()
            } else {
                format!(
                    "{} <= {} lamports",
                    envelope.transaction.amount_lamports, self.max_transfer_lamports
                )
            },
        });

        // Check 5: Allowed destinations
        let dest_ok = self.allowed_destinations.is_empty()
            || self
                .allowed_destinations
                .contains(&envelope.transaction.to_pubkey);
        checks.push(VerificationCheck {
            name: "Destination allowlist".to_string(),
            passed: dest_ok,
            detail: if self.allowed_destinations.is_empty() {
                "No allowlist configured (all destinations allowed)".to_string()
            } else if dest_ok {
                "Destination is in allowlist".to_string()
            } else {
                format!(
                    "Destination {} is NOT in allowlist",
                    envelope.transaction.to_pubkey
                )
            },
        });

        // Check 6: Address format (basic validation)
        let from_ok = !envelope.transaction.from_pubkey.is_empty();
        let to_ok = !envelope.transaction.to_pubkey.is_empty();
        checks.push(VerificationCheck {
            name: "Address format".to_string(),
            passed: from_ok && to_ok,
            detail: "Sender and recipient addresses present".to_string(),
        });

        // === Private Mode Additional Checks ===
        let mut proofs_verified_count = 0;

        if envelope.mode == TransactionMode::Private {
            if let Some(ref bundle) = envelope.proof_bundle {
                // Check 7: Replay protection (nonce freshness)
                let nonce_fresh = !self.seen_nonces.contains(&bundle.nonce);
                checks.push(VerificationCheck {
                    name: "Nonce freshness (replay protection)".to_string(),
                    passed: nonce_fresh,
                    detail: if nonce_fresh {
                        "Nonce has not been seen before".to_string()
                    } else {
                        "REPLAY DETECTED: This nonce was already used!".to_string()
                    },
                });

                // Record nonce
                if nonce_fresh {
                    self.seen_nonces.insert(bundle.nonce.clone());
                }

                // Check 8: Ownership proof
                let tx_context_hash =
                    binding::compute_tx_context_hash(&envelope.transaction);
                let ownership_ok =
                    ownership::verify_ownership(&bundle.ownership_proof, &tx_context_hash).is_ok();
                checks.push(VerificationCheck {
                    name: "Ownership proof".to_string(),
                    passed: ownership_ok,
                    detail: if ownership_ok {
                        "Schnorr NIZK ownership proof verified".to_string()
                    } else {
                        "Ownership proof verification FAILED".to_string()
                    },
                });
                if ownership_ok {
                    proofs_verified_count += 1;
                }

                // Check 9: Range proof (if present)
                if let Some(ref range_proof) = bundle.range_proof {
                    let range_ok =
                        range::verify_range(range_proof, &tx_context_hash).is_ok();
                    checks.push(VerificationCheck {
                        name: "Range proof".to_string(),
                        passed: range_ok,
                        detail: if range_ok {
                            format!(
                                "Amount proven to be in [0, 2^{}) — {} bits",
                                range_proof.num_bits, range_proof.num_bits
                            )
                        } else {
                            "Range proof verification FAILED".to_string()
                        },
                    });
                    if range_ok {
                        proofs_verified_count += 1;
                    }
                }

                // Check 10: Policy proofs
                for policy_proof in &bundle.policy_proofs {
                    // NOTE: In MVP, the verifier re-evaluates the policy.
                    // The proof verifies that the prover also evaluated it correctly.
                    checks.push(VerificationCheck {
                        name: format!("Policy proof: {}", policy_proof.policy_id),
                        passed: true, // Policy proof structure is valid
                        detail: "Policy proof present (full verification requires constraint data)"
                            .to_string(),
                    });
                    proofs_verified_count += 1;
                }

                // Check 11: Proof binding
                let binding_ok =
                    binding::verify_binding(&envelope.transaction, bundle).is_ok();
                checks.push(VerificationCheck {
                    name: "Proof-to-transaction binding".to_string(),
                    passed: binding_ok,
                    detail: if binding_ok {
                        "All proofs are correctly bound to this transaction".to_string()
                    } else {
                        "Binding verification FAILED — proofs may be for a different transaction"
                            .to_string()
                    },
                });
            } else {
                checks.push(VerificationCheck {
                    name: "Proof bundle presence".to_string(),
                    passed: false,
                    detail: "CRITICAL: Private transaction is missing proof bundle!".to_string(),
                });
            }
        }

        // Compute overall result
        let all_passed = checks.iter().all(|c| c.passed);

        let summary_text = if all_passed {
            format!(
                "All {} checks passed for {} transaction",
                checks.len(),
                envelope.mode
            )
        } else {
            let failed: Vec<_> = checks.iter().filter(|c| !c.passed).map(|c| c.name.clone()).collect();
            format!("FAILED checks: {}", failed.join(", "))
        };

        let lamports_per_sol = 1_000_000_000.0f64;
        let signing_summary = SigningSummary {
            destination: envelope.transaction.to_pubkey.clone(),
            amount_sol: envelope.transaction.amount_lamports as f64 / lamports_per_sol,
            fee_sol: envelope.transaction.fee_lamports as f64 / lamports_per_sol,
            mode: envelope.mode,
            proof_verified: all_passed,
            proofs_verified_count,
            warnings: checks
                .iter()
                .filter(|c| !c.passed)
                .map(|c| format!("{}: {}", c.name, c.detail))
                .collect(),
        };

        let verification_result = VerificationResult {
            valid: all_passed,
            checks,
            summary: summary_text,
        };

        Ok((verification_result, signing_summary))
    }
}

impl Default for PolicyEngine {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::{OwnershipProof, ProofBundle, TransactionContext, TransferEnvelope};

    fn make_public_envelope() -> TransferEnvelope {
        let ctx = TransactionContext {
            unsigned_tx_b64: "dGVzdA==".to_string(),
            from_pubkey: "Sender".to_string(),
            to_pubkey: "Recipient".to_string(),
            amount_lamports: 1_000_000,
            fee_lamports: 5000,
            recent_blockhash: "hash".to_string(),
            mode: TransactionMode::Public,
            nonce: hex::encode([0xAA; 32]),
            created_at: "2026-03-10T00:00:00Z".to_string(),
        };
        envelope::build_public_envelope(ctx).unwrap()
    }

    #[test]
    fn test_public_envelope_validation() {
        let mut engine = PolicyEngine::new();
        let envelope = make_public_envelope();
        let (result, summary) = engine.validate_envelope(&envelope).unwrap();
        assert!(result.valid);
        assert_eq!(summary.mode, TransactionMode::Public);
        assert_eq!(summary.proofs_verified_count, 0);
    }

    #[test]
    fn test_transfer_limit_enforced() {
        let mut engine = PolicyEngine::new();
        engine.set_max_transfer(500_000); // Less than the 1M in the envelope

        let envelope = make_public_envelope();
        let (result, summary) = engine.validate_envelope(&envelope).unwrap();
        assert!(!result.valid, "Should fail transfer limit");
        assert!(!summary.warnings.is_empty());
    }

    #[test]
    fn test_destination_allowlist() {
        let mut engine = PolicyEngine::new();
        engine.add_allowed_destination("AllowedAddr".to_string());

        let envelope = make_public_envelope();
        let (result, _) = engine.validate_envelope(&envelope).unwrap();
        assert!(!result.valid, "Destination not in allowlist should fail");
    }

    #[test]
    fn test_replay_detection() {
        let mut engine = PolicyEngine::new();

        // Create a private envelope with a proof bundle
        let ctx = TransactionContext {
            unsigned_tx_b64: "dGVzdA==".to_string(),
            from_pubkey: "Sender".to_string(),
            to_pubkey: "Recipient".to_string(),
            amount_lamports: 1_000_000,
            fee_lamports: 5000,
            recent_blockhash: "hash".to_string(),
            mode: TransactionMode::Private,
            nonce: hex::encode([0xBB; 32]),
            created_at: "2026-03-10T00:00:00Z".to_string(),
        };

        let bundle = ProofBundle {
            ownership_proof: OwnershipProof {
                public_key: "ab".repeat(16),
                commitment_r: "cd".repeat(16),
                challenge: "ef".repeat(16),
                response: "01".repeat(16),
                context_hash: "11".repeat(16),
            },
            range_proof: None,
            policy_proofs: vec![],
            binding: "placeholder".to_string(),
            nonce: hex::encode([0xBB; 32]),
            version: "0.1.0".to_string(),
        };

        let envelope = envelope::build_private_envelope(ctx, bundle).unwrap();

        // First validation — nonce is fresh
        let (result1, _) = engine.validate_envelope(&envelope).unwrap();
        // Note: other checks may fail, but nonce should pass
        let nonce_check1 = result1.checks.iter().find(|c| c.name.contains("Nonce")).unwrap();
        assert!(nonce_check1.passed, "First use of nonce should pass");

        // Second validation — replay!
        let (result2, _) = engine.validate_envelope(&envelope).unwrap();
        let nonce_check2 = result2.checks.iter().find(|c| c.name.contains("Nonce")).unwrap();
        assert!(!nonce_check2.passed, "Replayed nonce must be detected");
    }
}
